### PR TITLE
Pipeop colroles

### LINKED ISSUE #273: PipeOp for setting / changing col roles
e.g. add / remove stratification, weights, blocking, ...

### PR DESCRIPTION
**Pull request by sumny**
closes #273 
Setting a new `target` or changing the role of an existing `target` is not supported. #139 will allow for this (?). Col roles are specified via the parameter `new_role` (list) that must be named with the names matching the column of the task that will be used later for training/prediction. Elements of `new_role` must be character vectors of `"feature"`, `"name"`, `"order"`, `"group"`, `"stratum"`,  and `"weight"`. If the value is `character()`, the column will be dropped from the task. Maybe it would be nice to just set a column to `"unused"` (so we should think about implementing this `col_role` type in general?).

### CODE REVIEW

File: R/PipeOpColRoles.R [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    },
+
+    transform = function(task) {
+      # names of "new_role" must be a subset of the column names of the task
+      assert_subset(names(self$param_set$values$new_role), choices = task$col_info$id)
+      # changing the role of a target is not supported
+      if (any(task$col_roles$target %in% names(self$param_set$values$new_role))) {
+        stop("Cannot change the role of a target.")
+      }
+      col_roles = unlist(task$col_roles)
```
[the remark this answers is not in the document]
Comment (pfistfl) [reply R/PipeOpColRoles.R]: This should remove old rows automatically etc.

File: R/PipeOpColRoles.R [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+          # a value of "character()" will lead to this column being dropped
+          check_subset(unlist(x), c("feature", "name", "order", "group", "stratum", "weight"))
+        })
+      ))
+      super$initialize(id, param_set = ps, param_vals = param_vals, can_subset_cols = FALSE)
+    },
+
+    transform = function(task) {
+      # names of "new_role" must be a subset of the column names of the task
+      assert_subset(names(self$param_set$values$new_role), choices = task$col_info$id)
```
Comment (pfistfl) [inline R/PipeOpColRoles.R]: save `self$param_set$values$new_role` into a vector once, you access this super long name really often.

File: R/PipeOpColRoles.R [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+#'
+#' @section State:
+#' The `$state` is a named `list` with the `$state` elements inherited from [`PipeOpTaskPreproc`].
+#'
+#' @section Parameters:
+#' The parameters are the parameters inherited from [`PipeOpTaskPreproc`], as well as:
+#' * `new_role`  :: `list` \cr
+#'   Named list of new column roles. The names must match the column names of the input task that
+#'   will later be trained on. Each entry of the list must contain a character vector with possible
+#'   values of `feature`, `name`, `order`, `group`, `stratum`,  and `weight`. If the value is given
```
[the remark this answers is not in the document]
Comment (sumny) [reply R/PipeOpColRoles.R]: thanks! wasn't aware of that.

File: R/PipeOpColRoles.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (sumny) [reply R/PipeOpColRoles.R]: Can you specify what exactly you mean with `old` rows?

**Comment by sumny** [comment]
> I think you might be able to heavily simplify this using `set_col_role` instead.

thanks for the code review! I also thought about using `set_col_role`, however, in `?Task` it is stated that:

Method 'set_col_role()':
..
This function is deprecated and will be removed in the next
version in favor of directly modifying $col_roles.

So I thought I should not use it.

File: R/PipeOpColRoles.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (pfistfl) [reply R/PipeOpColRoles.R]: old roles*, sorry. 
Later on you remove the old roles for some cols and add new ones, this is not neccessary.

**Review by pfistfl — APPROVED** [review]
LGTM

**Pull request merged**